contracts/BuilderStaking.sol

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {AccessControlDefaultAdminRulesUpgradeable} from '@openzeppelin/contracts-upgradeable/access/extensions/AccessControlDefaultAdminRulesUpgradeable.sol';
+import {UUPSUpgradeable} from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+import {Initializable} from '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import {SafeERC20} from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+/**
+ * @title Divvi BuilderStaking
+ * @dev Contract for managing $DIVVI token stakes on behalf of builders
+ * @custom:security-contact security@valora.xyz
+ */
+contract BuilderStaking is
+  Initializable,
+  AccessControlDefaultAdminRulesUpgradeable,
+  UUPSUpgradeable
+{
+  using SafeERC20 for IERC20;
+
+  // State variables
+  IERC20 public divviToken;
+  uint256 public stakingThreshold;
+  uint256 public totalStaked;
+
+  mapping(address => mapping(address => uint256)) public stakersForBeneficiary; // beneficiary => staker => amount
+  mapping(address => mapping(address => uint256)) public beneficiariesForStaker; // staker => beneficiary => amount
+
+  // Arrays to track relationships for efficient querying
+  mapping(address => address[]) public stakerListForBeneficiary;
+  mapping(address => address[]) public beneficiaryListForStaker;
+
+  // Events
+  event ThresholdUpdated(uint256 newThreshold, uint256 previousThreshold);
+  event Staked(
+    address indexed staker,
+    address indexed beneficiary,
+    uint256 amount
+  );
+  event Unstaked(
+    address indexed staker,
+    address indexed beneficiary,
+    uint256 amount
+  );
+
+  // Errors
+  error ZeroAddressNotAllowed();
+  error AmountMustBeGreaterThanZero();
+  error InsufficientStakeBalance(uint256 requested, uint256 available);
+  error CannotRescueStakedTokens(uint256 requested, uint256 totalStaked);
+  error CannotRescueZeroAmount();
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  /**
+   * @dev Initializes the contract
+   * @param _divviToken Address of the $DIVVI token contract
+   * @param _admin Address that will have DEFAULT_ADMIN_ROLE
+   * @param _initialThreshold Initial staking threshold (can be 0)
+   */
+  function initialize(
+    address _divviToken,
+    address _admin,
+    uint256 _initialThreshold
+  ) external initializer {
+    if (_divviToken == address(0)) revert ZeroAddressNotAllowed();
+    if (_admin == address(0)) revert ZeroAddressNotAllowed();
+
+    __AccessControlDefaultAdminRules_init(0, _admin);
```

Comment:
Should we set the default transfer delay to something higher?

Reply:
Sure -- can set it to something like one day as a default.

Reply:
Actually, I've updated this to be a constructor param, in line with what we do in other contracts.